### PR TITLE
Specify TreeDB rabitq_1bit v1 reference codec

### LINKED ISSUE #2449: Specify TreeDB rabitq_1bit v1 codec and pure-Go reference kernels
Parent tracker: https://github.com/snissn/gomap/issues/2447

## Goal

Define TreeDB's `rabitq_1bit` v1 codec contract and add a pure-Go reference implementation that can serve as the correctness oracle for all durable asset and acceleration work.

## Scope

- Define codec name/version/config identity, including deterministic seed/rotation/query encoding choices.
- Decide v1 code storage shape:
  - `packed_codes` as `packed_bit_vector` with TreeDB LSB-first bit order, or
  - `codes` as `uint64_vector` with go-highway-compatible word order.
- Document vector dimensions vs code dimensions and `CodeWidthBits` for the selected shape.
- Add `TreeDB/internal/rabitq` or equivalent pure-Go reference code for:
  - data-vector encoding;
  - query encoding / weighted-popcount inputs;
  - side-array computation such as `code_count` and `quantized_dot_product_inv`;
  - estimator scoring for the first supported metric.
- Support cosine / normalized column_graph first. Other metrics can be deferred.
- Include golden tests for deterministic output, bit order, padding, zero/degenerate vectors, and estimator sanity.

## Non-Goals

- No typed-column asset publication in this issue.
- No collection/search integration.
- No SIMD/go-highway backend yet.
- No multi-bit RaBitQ, PQ/OPQ, or IVF work.

## Required Tests / Evidence

- [ ] Golden vectors for encode/query/score with stable seed/config.
- [ ] Bit-order and padding tests for the selected storage shape.
- [ ] Pure-Go scorer sanity versus exact cosine on small deterministic fixtures.
- [ ] Benchmark micro-row for reference scorer if useful, explicitly not a production speed claim.

## Completion Criteria

- [ ] `rabitq_1bit` v1 codec contract is documented enough for durable assets.
- [ ] Pure-Go implementation is allocation-disciplined and usable as an oracle.
- [ ] Later asset/scorer PRs can depend on this without reopening storage-shape decisions.

### PR DESCRIPTION
**Pull request by snissn**
## Objective

Specify TreeDB's `rabitq_1bit` v1 codec contract and add a clean-room pure-Go reference oracle for encode/query/score behavior.

Closes #2449. Parent tracker: #2447.

## Context

This is the design/contract gate for later RaBitQ asset persistence and scorer integration. TreeDB already has packed-bit typed-column storage and quantized asset roles; this PR chooses the v1 code storage/API contract and provides deterministic reference kernels.

## Non-goals

- No typed-column asset publication.
- No collection/search integration.
- No SIMD/go-highway backend.
- No multi-bit RaBitQ/PQ/OPQ/IVF work.
- No scalar_u8 or exact/default FP32 behavior changes.
- No go.mod/toolchain changes.

## Design

- Adds `TreeDB/internal/rabitq`:
  - `CodecName = "rabitq_1bit"`, `CodecVersion = 1`, `DefaultSeed = 0x7261626974710001`.
  - Canonical config bytes/hash for future manifest identity.
  - `NewPlan`, `Encode`, `EncodeQuery`, `ScoreCosine`/`ScoreEncoded`, `ValidateCode`, `ValidateQuery`.
- Storage contract selected:
  - `quantizedasset.RolePackedCodes` / `packed_codes`.
  - typed-column `packed_bit_vector` / `raw_packed_bit_vector`.
  - TreeDB LSB-first bit order, zero high-bit padding.
  - `CodeDimensions = next_power_of_two(VectorDimensions)`, `CodeWidthBits = 1`.
- Reference rotation/scoring:
  - deterministic signed permutation + normalized FWHT over padded dims;
  - query sign bits + absolute weights as weighted-popcount inputs;
  - side arrays: `code_count`, `quantized_dot_product_inv`;
  - estimated cosine documented in `TreeDB/docs/spec/rabitq-1bit-v1.md`.

## Scope

Includes:
- `TreeDB/internal/rabitq/*`
- `TreeDB/docs/spec/rabitq-1bit-v1.md`
- docs links/tests in `TreeDB/docs/*` and quantized specs.

Excludes:
- `TreeDB/collections` implementation changes.
- quantized asset publication/search/scorer integration files.
- go-highway/SIMD and go.mod/go.sum.

## Correctness

Tests run:
- `go test ./TreeDB/internal/rabitq ./TreeDB/docs ./TreeDB/internal/quantizedasset ./TreeDB/internal/typedcolumn`
- `go test ./TreeDB/collections -run 'Test(VectorIndexQuantizedDefinitionNormalization1926|ColumnGraphScalarU8QuantizedAssetRebuildPrepareReopen1926|ColumnGraphScalarU8QuantizedOnlyUsesPreparedCodes1926|ColumnGraphScalarU8CenteredQueryScratch2258)'`

Final base-sync validation after #2458 merge (`82bf8754d`):
- `GOWORK=off go test ./TreeDB/internal/rabitq ./TreeDB/docs ./TreeDB/internal/quantizedasset ./TreeDB/internal/typedcolumn -count=1`
- `GOWORK=off go test ./TreeDB/collections -run 'Test(VectorIndexQuantizedDefinitionNormalization1926|ColumnGraphScalarU8QuantizedAssetRebuildPrepareReopen1926|ColumnGraphScalarU8QuantizedOnlyUsesPreparedCodes1926|ColumnGraphScalarU8CenteredQueryScratch2258|GoHighwayRaBitQSmoke)' -count=1`
- Latest-head CI is green after rerunning one flaky/unrelated `TreeDB/mongo_gateway` Linux job (`TestStandaloneServerCommandWALBSONInsertReadUpdateVisibility/command_wal_relaxed`).

Coverage added:
- golden encode/query/score with stable seed/config;
- packed LSB-first bit order and padding validation;
- zero/non-finite/dimension mismatch fail-closed behavior;
- malformed scorer side-input fail-closed behavior;
- scorer sanity versus exact cosine fixtures;
- zero steady-state allocations after warmup.

Internal review:
- High-thinking reviewer subagent PASS at `bf64b0ece`; blockers fixed and rechecked.

## Performance

Reference-only microbenchmark (not a production speed claim):

Command:
- `go test ./TreeDB/internal/rabitq -run '^$' -bench '^BenchmarkReferenceScoreCosine2449$' -benchmem -count=3`

Apple M3 / darwin arm64:

| Row | ns/op | MB/s | B/op | allocs/op |
| --- | ---: | ---: | ---: | ---: |
| run 1 | 298.5 | 53.61 | 0 | 0 |
| run 2 | 312.8 | 51.14 | 0 | 0 |
| run 3 | 342.4 | 46.73 | 0 | 0 |

Regression status: no material hot-path regression expected; this PR adds a new internal reference package and docs, with no collection/search/scalar_u8 path changes.

## Rollout / Toggle Plan

No runtime behavior is enabled by this PR. Later asset/scorer PRs must explicitly integrate against this contract and fail closed on mismatches.

## Follow-ups

- #2450 durable RaBitQ quantized assets.
- #2451 lower-level RaBitQ scoring.
- #2452 collection-level buffered support.
- #2453 go-highway acceleration.
- #2454 recall/storage/performance/docs closeout.

## AI Review Loop

- Codex: first review on `bf64b0ece` found one side-input validation issue; fixed in `cac9263cd`, thread resolved, and latest-head re-review on `82bf8754d` reported no major issues.
- Copilot: requested on latest head; no Copilot review/comment was produced in this repo, so no Copilot findings are outstanding.
- CodeRabbit: latest-head/incremental review completed; status check is passing and no inline review comments/threads are outstanding.
- Review threads: 1 resolved / 0 unresolved.

### CODE REVIEW

**Comment by snissn** [comment]
@copilot review

**Comment by snissn** [comment]
@codex review

**Comment by snissn** [comment]
@copilot review

**Comment by snissn** [comment]
@codex review

File: TreeDB/internal/rabitq/rabitq.go [hunk 1]
```diff
@@ -0,0 +1,452 @@
...
+func (p *Plan) ScoreCosine(query Query, code []byte, codeCount uint32, quantizedDotProductInv float32) (float64, error) {
+	if err := p.ValidateCode(code, codeCount); err != nil {
+		return 0, err
+	}
+	if err := p.ValidateQuery(query); err != nil {
+		return 0, err
+	}
+	if quantizedDotProductInv <= 0 || math.IsNaN(float64(quantizedDotProductInv)) || math.IsInf(float64(quantizedDotProductInv), 0) {
+		return 0, fmt.Errorf("%w: invalid quantized_dot_product_inv=%v", ErrDegenerateVector, quantizedDotProductInv)
+	}
```
[the remark this answers is not in the document]
Comment (snissn) [reply TreeDB/internal/rabitq/rabitq.go]: Fixed in cac9263cd: ScoreCosine now calls ValidateQuantizedDotProductInv, which rejects non-finite values and finite values outside the unit-L2 rotated range [1/sqrt(CodeDimensions), 1] with rounding tolerance; tests cover tiny and >1 qdp_inv values.

**Comment by snissn** [comment]
@coderabbitai review

**Comment by snissn** [comment]
@codex review

**Comment by snissn** [comment]
@copilot review

**Comment by snissn** [comment]
@coderabbitai review

**Comment by snissn** [comment]
## Final status for #2449

Latest PR head before merge: `82bf8754d61e10430b4d05d201b276d9e9e99930`.
Merge commit: `7681e85483cb2d0f1f1107b3e9d6d6c8aa40eedd`.

### Implemented scope

- Added `TreeDB/internal/rabitq` pure-Go reference/oracle package for `rabitq_1bit` v1.
- Documented the codec contract in `TreeDB/docs/spec/rabitq-1bit-v1.md` and linked it from quantized specs.
- Storage/API contract chosen:
  - codec `rabitq_1bit`, version `1`;
  - canonical config identity with default seed `0x7261626974710001`;
  - deterministic signed-permutation + normalized FWHT rotation, fully specified constants/order;
  - `RolePackedCodes` / `packed_bit_vector` / `raw_packed_bit_vector`;
  - TreeDB LSB-first bit order with zero high-bit padding;
  - `CodeDimensions = next_power_of_two(VectorDimensions)`, `CodeWidthBits = 1`;
  - side arrays `code_count` and `quantized_dot_product_inv`;
  - fail-closed validation for code rows, query side inputs, and `quantized_dot_product_inv` range.

### Non-goals preserved

No typed-column asset publication, collection/search integration, SIMD/go-highway backend, scalar_u8/exact behavior change, multi-bit/PQ/OPQ/IVF work, or go.mod/toolchain ownership in this PR.

### Evidence

Local tests:

- `go test ./TreeDB/internal/rabitq ./TreeDB/docs ./TreeDB/internal/quantizedasset ./TreeDB/internal/typedcolumn`
- `go test ./TreeDB/collections -run 'Test(VectorIndexQuantizedDefinitionNormalization1926|ColumnGraphScalarU8QuantizedAssetRebuildPrepareReopen1926|ColumnGraphScalarU8QuantizedOnlyUsesPreparedCodes1926|ColumnGraphScalarU8CenteredQueryScratch2258)'`

Reference-only microbenchmark (not production search speed):

- `go test ./TreeDB/internal/rabitq -run '^$' -bench '^BenchmarkReferenceScoreCosine2449$' -benchmem -count=3`

Latest local run on Apple M3 / darwin arm64:

| run | ns/op | MB/s | B/op | allocs/op |
| --- | ---: | ---: | ---: | ---: |
| 1 | 412.2 | 38.82 | 0 | 0 |
| 2 | 484.9 | 33.00 | 0 | 0 |
| 3 | 620.0 | 25.81 | 0 | 0 |

Performance regression status: no material hot-path regression; new internal reference package/docs only, no collection/search/scalar_u8 path changes.

### Review / CI

- Internal high-thinking review: PASS after qdp range fix; no blockers/nits remaining.
- Codex: initial qdp_inv range finding fixed in `cac9263cd`; thread resolved; latest Codex comment reported no major issues.
- CodeRabbit: requested after maturity; status check passed/skipped with no actionable review threads.
- Copilot: requested; no bot response observed.
- Latest-head CI for `82bf8754d61e10430b4d05d201b276d9e9e99930`: green before merge.

#2450 can safely build on the merged contract.
